### ISSUE TITLE
吃瓜教程第5章

### ISSUE THREAD

**Issue by Edwlr** [Edwlr/eatingmelon#4]
# 第5章：神经网络

神经网络是由具有适应性的简单单元组成的广泛并行互连的网络，它的组织能够模拟生物神经系统对真实世界物体所作出的交互反应。
神经网络中最基本的成分是神经元模型,将多个神经元按一定层次结构连接起来，就得到神经网络。
使用深层神经网络类算法所做的机器学习为深度学习。

M-P神经元模型：神经元接收到来自 $n$个其他神经元传递过来的输入信号，这些输入信号通过带权重的连接进行传递，神经元接收到的总输入值将与神经元的阈值进行比较，然后通过激活函数处理以产生神经元输出。
[](url)
理想的激活函数是阶跃函数 $\mathrm{sgn}(x)$，但由于其不连续、不光滑，实际一般采用Sigmoid函数 $\mathrm{sigmoid}(x)=\frac{1}{1+\mathrm{e}^{-x}}$。

### 感知机(Perceptron)
感知机由两层神经元组成，输入层接收外界信号传递给输出层，输出层是M-P神经元，也称为阈值逻辑单元(threshold logic unit)
感知机能很容易实现与、或、非的逻辑运算
感知机模型可表示为 $$y=f(w^\mathrm{T}x-\theta)$$ $x\in \mathbb{R}^n$为样本的特征向量，是感知机模型的输入； $w,\theta$是感知机模型的参数。
$w^\mathrm{T}x-\theta$是 $n$维空间一个超平面，当采用阶跃函数作为激活函数时，超平面将空间划分为 $w^\mathrm{T}x-\theta\ge0$和 $w^\mathrm{T}x-\theta<0$两个子空间，从而实现分类的功能。
给定一个线性可分的数据集 $T$，感知机的学习目标是求得能对 $T$中正负样本完全正确划分的超平面 $w^\mathrm{T}x-\theta=0$。假设误分类的样本集合为 $M\subseteq T$，对任意一个误分类的样本 $(x,y)\in M$，有 $w^\mathrm{T}x-\theta\ge0$时，模型输出 $\hat{y}=1$， 样本真实标记为 $y=0$；反之，$w^\mathrm{T}x-\theta<0$时，模型输出 $\hat{y}=0$， 样本真实标记为 $y=1$。故有 $$(\hat{y}-y)(w^\mathrm{T}x-\theta)\ge0$$。从而对于数据集 $T$，损失函数可定义为 $$L(w,\theta)=\sum_{x\in M}(\hat{y}-y)(w^\mathrm{T}x-\theta)$$感知机模型的学习问题就可以转化为极小化损失函数的问题。
将阈值 $\theta$看作一个固定输入为-1的哑结点(dummy node)所对应的连接权值 $w_{n+1}$，则损失函数简化为 $L(w)=\sum_{x\in M}(\hat{y}-y)w^\mathrm{T}x$，其梯度为 $\nabla_w L(w)=\sum_{x\in M}(\hat{y}-y)x$
感知机的学习算法采用随机梯度下降法，故权重 $w$的更新公式为 $$w_i\leftarrow w_i+\Delta w_i=w_i+\eta(y-\hat{y})x_i$$其中 $\eta\in(0,1)$为学习率。
感知机只有一层功能神经元，难以甚至不能解决非线性可分问题。解决非线性可分问题，需要考虑使用多层功能神经元。

### BP神经网络
多层前馈网络(multi-layer feedforward neural network)：每层神经元与下一层神经元全互连，神经元之间不存在同层连接，也不存在跨层连接。
输出层与输入层之间的一层神经元被称为隐含层，隐含层是拥有激活函数的功能神经元。
只需一个包含足够多神经元地隐含层，多层前馈网络就能以任意精度逼近任意复杂度的连续函数。
误差逆传播(error BackPropagation，简称BP)算法是训练多层网络中最成功的学习算法。BP算法不仅可用于多层前馈神经网络，还可用于其他类型的神经网络。但通常说的BP网络一般指用BP算法训练的多层前馈神经网络。
[](url)
给定训练集 $D=\\{(x_1,y_1),\dots,(x_m,y_m)\\},x_i\in\mathbb{R}^d,y_i\in\mathbb{R}^l$，即输入示例由 $d$个属性描述，输出 $l$维实值向量。对训练例 $(x_k,y_k)$，神经网络的输出为 $\hat{y}_k=(\hat{y}_1^k,\dots,\hat{y}_l^k)$，则网络在 $(x_k,y_k)$上的均方误差为 $$E_k=\frac{1}{2}\sum_{j=1}^l(\hat{y}_j^k-y_j^k)^2$$ 
BP算法的参数更新与感知机学习规则类似，参数 $v$的更新式为 $v\leftarrow v+\Delta v=v-\eta\nabla_vE$， $E$为损失函数， $\eta\in(0,1)$为学习率。
标准BP算法每次仅针对一个训练样本更新连接权重和阈值。若类似地推导基于累积误差最小化地更新规则，则能得到累积BP算法。
缓解BP网络过拟合方法：
- 早停
  将数据分为训练集和验证集，训练集用于计算梯度、更新连接权和阈值，验证集用于估计误差，若训练集误差降低而验证集误差升高，则停止。
- 正则化
  在误差目标函数中增加一个用于描述网络复杂度的部分，如连接权与阈值的平方和。

### 其他神经网络
- RBF(Radial Basis Function)网络
  单隐层前馈神经网络，使用径向基函数作为隐层神经元激活函数，输出层为对隐层神经元输出的线性组合
- ART(Adaptive Resonance Theory)网络
  竞争型学习是神经网络中一种常用的无监督学习策略，网络的输出神经元相互竞争，每一时刻仅有一个竞争获胜的神经元被激活，其他神经元被抑制。
  ART为竞争型学习的重要代表。网络由比较层、识别层、识别阈值和重置模块构成。ART可进行增量学习或在线学习。
- SOM(Self-Organizing Map)网络
  竞争学习型的无监督神经网络，能将高维输入数据映射到低维空间，同时保持输入数据在高维空间的拓扑结构。
- 级联相关(Cascade-Correlation)网络
  一般的神经网络模型通常事先固定网络结构，结构自适应网络则将网络结构也作为学习目标之一。
  级联相关网络是结构自适应网络的重要代表。与一般前馈神经网络相比，级联相关网络无需设置网络层数、隐含层神经元数目，且训练速度较快，但在数据较小时任意过拟合。
- Elman网络
  递归神经网络(recurrent neural networks)允许网络中出现环状结构，从而让一些神经元的输出反馈回来作为输入信号。
  Elman网络是最常用的递归神经网络之一。隐层神经元的输出被反馈回来与下一时刻输入层的神经元信号一起作为下一时刻的输入。隐含层神经元通常采用Sigmoid激活函数，网络训练采用推广的BP算法。
- Boltzmann机
  神经网络中有一类模型为网络状态定义一个能量，能量最小化时网络达到理想状态。Boltzmann机就是一种基于能量的模型。
  若网络中的神经元以任意不依赖于输入值的顺序进行更新，则网络最终将达到Boltzmann分布。
